### PR TITLE
String documentation improvements

### PR DESCRIPTION
**Pull request by dbuenzli**
The `Stdlib.String` module always makes me sad. This is a first step at trying to make myself less sad. I also hope it can provide a clearer view on how to evolve the design of the module to make me even more less sad.

It is purely a documentation PR. It rewrites the module preamble, reorders and sections the functions, tweaks and uniformizes the doc strings. It's not perfect yet but I have timed out on the time I decided to allocate for it and it's still a good improvement in my opinion.

I suspect the diff is difficult to follow so if you prefer looking at the rendered documentation you can have a look at the  [old rendering](https://erratique.ch/tmp/ocaml-string/old/tmp/String/) and the  [new rendering](https://erratique.ch/tmp/ocaml-string/new/tmp/String/).

I also want to warn upfront that I sneaked `String.blit` under deprecated functions sections while it's not marked as such formally – I didn't want to introduce a section for it and I think Bytes.blit_string is fine and should be used instead.

### CODE REVIEW

**Review by gasche — APPROVED** [review]
I like the changes, I believe they are going in the right direction. I'm approving, but as far as stdlib changes go I would prefer to have a second pair of reviewing eyes before making merging decisions.

There is a lot of categorization, and the rewordings are usually very minor (I didn't really find anything in there that someone could object to). The change to the preamble is a bigger change -- if you want to check that there is nothing you dislike in this PR, you should probably start there. In particular, it carefully defines distinct notions of "indices" and "positions" in a string.

I have one nitpick: you renamed the integer parameters of `make` and `init` from `n` to `l` (like `length`, I suppose), and I would prefer to see the `n` back. It (comes from mathematical tradition, but this we don't really want to admit, and) is used consistently in Array, Bytes etc. for that purpose, and is also a common naming choice for sizes/lengths in many other communities and languages. (On the other hand I support the move away from `n`for indices, typically to `i`.)

**Comment by dbuenzli** [comment]
> I have one nitpick: you renamed the integer parameters of `make` and `init` from `n` to `l` (like `length`, I suppose), and I would prefer to see the `n` back.

I don't have a strong opinion on that, I can do it. (Note however that `l` was already used there before in the preamble).

**Comment by Octachron** [comment]
I note that `StringLabels` have not been updated. I would thus suggest to wait for the unified documentation PR (#9745) before merging this PR (#9745 is on my review list for this week).

File: stdlib/string.mli
Comment (xavierleroy) [inline stdlib/string.mli]: referred

File: stdlib/string.mli
Comment (xavierleroy) [inline stdlib/string.mli:42]: [i]-th ?

File: stdlib/string.mli
Comment (xavierleroy) [inline stdlib/string.mli]: Run-on sentence.  What about: "indices and positions.  These are depicted and described as follows." ?

File: stdlib/string.mli
Comment (xavierleroy) [inline stdlib/string.mli]: "It represents either" ?   The 3-case enumeration is exhaustive.

**Review by xavierleroy — COMMENTED** [review]
I like the grouping in sections and the ordering of these sections, starting with the most commonly used functions.  This discipline (grouping + order) should be applied to other standard library modules such as List and Array, which, currently, are a bit of a mess.

I support using `n` in preference to `l` to denote the number of characters of a string, if only because `n` is more visually distinctive, while `l` can be confused for `I` or `1`.

I spotted one typo and suggested alternate wordings below.  It would be nice to have feedback from a native English speaker.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli:42]: Cambridge says [`nth`](https://dictionary.cambridge.org/dictionary/english/nth). So I'll keep `ith`.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli]: Suggestion applied.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli]: Suggestion applied.

File: stdlib/string.mli
Comment (jberdine) [inline stdlib/string.mli]: Splitting hairs, but AFAIU better efficiency is only in the case when type-based optimizations don't trigger.
```suggestion
    behaves like {!Stdlib.compare} on strings but may be more efficient. *)
```

File: stdlib/string.mli
Comment (jberdine) [inline stdlib/string.mli]: It is perhaps worth consciously deciding whether "iff" is a widely-enough known abbreviation to use here.

File: stdlib/string.mli
Comment (jberdine) [inline stdlib/string.mli]: Perhaps:
```suggestion
    increasing order. In "unsafe-string" mode, modifications of the string during iteration
```

File: stdlib/string.mli
Comment (jberdine) [inline stdlib/string.mli]: I'm no linguistic authority, but to me it would flow better to prefer words over punctuation:
```suggestion
    A string [s] of length [n] is an indexable and immutable sequence of
```

**Review by jberdine — COMMENTED** [review]
For what it's worth, looks good to me.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli]: At worst people will think it's a typo.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli]: Good point. Changed as suggested.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli]: Definitively better. Changed as suggested.

File: stdlib/string.mli
Comment (dbuenzli) [reply stdlib/string.mli]: Reminds me I need a haircut. Changed as suggested.

**Comment by dbuenzli** [comment]
So this was rebased over #9745. Well redone (with the latest comments integrated) as it was impossible to resolve conflicts.

It seems that #9745 added a comment and instructions on how to use the labelled version of the stdlib. I think that this is not the place where it belongs and should be mentioned elsewhere (if at all) in the manual. The programmer after having understood that maybe she didn't need any of:

* extlib
* batteries
* base
* core
* core_kernel
* stringext
* astring
* ...

can be left in peace with her choice without the need to provide her with yet another alternative (not to mention that the example shown is a case of pointless labelling and a style that should be discouraged). If there is strong objection about this I will discard the commit.

**Comment by gasche** [comment]
In #9936 @dbuenzli asked about this PR. My answer is:
- I would like to merge the PR, which is a nice improvement to the documentation.
- I don't have an opinion about what to do on the paragraph that mentions labels, but I think this must be discussed with the author of the labelling-documentation change, @johnwhitington, if you want to change something.

Right now, no discussion has happened, so I cannot answer your question. My current understanding is that I should wait for the discussion to happen before merging. I would be happy to merge *right now* if we decide to do nothing in this PR that requires further synchronization. I think that means dropping your change suggestion for now, and inviting you to discuss it in a separate PR.

**Comment by xavierleroy** [comment]
> It seems that #9745 added a comment and instructions on how to use the labelled version of the stdlib. I think that this is not the place where it belongs and should be mentioned elsewhere (if at all) in the manual. 

I tend to agree: I would prefer more explanations in `stdLabels.mli` on how to use the stdlib with labels, and no explanations (or just a mention "see also `StdLabels.String`") in `string.mli`, `array.mli`, etc. That's a part of #9745 that escaped review, apparently.  So much back-and-forth in the stdlib is disturbing, however.

**Comment by johnwhitington** [comment]
Before #9745, `*Labels.mli` had the `StdLabels` instructions at the top. Unlabeled modules had no reference at all to labels.

After #9745, the same instructions appear at the top of both labeled and unlabeled modules.

If this is unacceptable, there are three options:

1) Remove all mention of StdLabels from both labeled and unlabled modules
2) Shorten the description "just a mention" as @xavierleroy says above. Still appears in both labeled and unlabeled
3) Modify `tools/sync_stdlib_docs` to remove the `StdLabels` instructions when generating the unlabeled modules

**Comment by johnwhitington** [comment]
In my opinion, any of options 1, 2, and 3 above should be in a separate PR, since they will touch Bytes, Array and List as well as String.

**Comment by Octachron** [comment]
Since removing the mention to the labelled version of module should be done uniformly in `String`, `Array`, `List` and `Bytes`. I think it is better to split this second commit, merge this PR without this change, and discuss independently this paragraph on the use of labelled modules.

**Comment by dbuenzli** [comment]
I really think these things belong to the manual and not the API docs, they just get in the way as they stand now. In particular chapter 4 should be opening `open Stdlabels` rather than using `StringLabels` and `ListLabels` directly.

**Comment by Octachron** [comment]
I tend to agree that the labelled version of modules do not need to be advertised as insistently as currently and that the option 1 is probably the best.

**Comment by dbuenzli** [comment]
(Offending commit removed, no changes entry needed)

**Comment by MisterDA** [comment]
> I also want to warn upfront that I sneaked `String.blit` under deprecated functions sections while it's not marked as such formally – I didn't want to introduce a section for it and I think `Bytes.blit_string` is fine and should be used instead.

The position of `String.blit` in the `String` module documentation is a bit weird now, just after the UTF conversion functions. I think it'd better be placed above these, and maybe with a link to `Bytes.blit_string` if that's preferred?

**Comment by dbuenzli** [comment]
Indeed it looks wrong. Not exactly sure what happened here.

**Pull request merged**
